### PR TITLE
PoC: WebSocket message handler.

### PR DESCRIPTION
**Pull request by Goldziher**
- Adds handler class `WSMessageHandler` and alias `ws_message`.
- reimplements `WebSocket.{send_json,receive_json}()` methods so we use `orjson`.
- Adds `WSMessageRoute` type - the socket connection is maintained in `handle()`

### CODE REVIEW

File: starlite/connection.py [hunk 1]
```diff
@@ -74,3 +75,32 @@ def auth(self) -> Auth:
     @property
     def query_params(self) -> Dict[str, Any]:  # type: ignore[override]
         return parse_query_params(self)
+
+    async def receive_json(self, mode: str = "text") -> Any:
+        """
+        Exact copy of the `starlette` method, but using `orjson.loads()`.
+        """
+        if mode not in {"text", "binary"}:
+            raise RuntimeError('The "mode" argument should be "text" or "binary".')
```
Comment (Goldziher) [inline starlite/connection.py:84]: Should probably be one of our exceptions or a new exception of ours that inherits from both StarliteException and RuntimeError.

File: starlite/handlers/__init__.py [hunk 2]
```diff
@@ -16,4 +18,5 @@
     "put",
     "route",
     "websocket",
+    "ws_message",
```
Comment (Goldziher) [inline starlite/handlers/__init__.py:21]: `wsmessage`?

File: starlite/handlers/websocket_message.py [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+    in `receive()` being called are not allowed to be called.
+
+    As many properties of the `HTTPRouteHandler` are supported, however they wrap the
+    socket connection, not each message.
+
+    A second set of properties that wrap the message are available.
+
+    Thinking
+    --------
+    - should I just use after_request and before_request as after_disconnect and before_accept?
```
Comment (Goldziher) [inline starlite/handlers/websocket_message.py:35]: its a good idea to have lifecycle hooks for these.

File: starlite/handlers/websocket_message.py [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+    """
+
+    __slots__ = (
+        "handler_fn",
+        "before_accept",
+        "after_disconnect",
```
Comment (Goldziher) [inline starlite/handlers/websocket_message.py:41]: maybe also a before_disconnect?

File: starlite/handlers/websocket_message.py [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+        "before_message",
+        "after_message",
+    )
+
+    @validate_arguments(config={"arbitrary_types_allowed": True})
+    def __init__(
+        self,
+        before_accept: Optional[Callable[[WebSocket], WebSocket]] = None,
+        after_disconnect: Optional[Callable[[WebSocket], WebSocket]] = None,
+        before_message: Optional[Callable[[Any], Any]] = None,
```
Comment (Goldziher) [inline starlite/handlers/websocket_message.py:51]: please use named types for these and place them under `starlite.types`

File: starlite/router.py [hunk 6]
```diff
@@ -96,7 +103,7 @@ def route_handler_method_map(self) -> Dict[str, Union[WebsocketRouteHandler, Dic
     @staticmethod
     def map_route_handlers(
         value: Union[Controller, BaseRouteHandler, "Router"],
-    ) -> ItemsView[str, Union[WebsocketRouteHandler, ASGIRoute, Dict[HttpMethod, HTTPRouteHandler]]]:
+    ) -> ItemsView[str, Union[WebsocketRouteHandler, WSMessageHandler, ASGIRoute, Dict[HttpMethod, HTTPRouteHandler]]]:
```
Comment (Goldziher) [inline starlite/router.py:106]: i think its time to introduce some type aliases for this 😉

File: starlite/routes.py [hunk 7]
```diff
@@ -304,6 +306,86 @@ async def handle(self, scope: Scope, receive: Receive, send: Send) -> None:
...
+        Holds the websocket connection and feeds parsed/validated messages into the handler.
+
+        - create WebSocket
+        - accept the connection after running guards
+        - iterate over the socket
+        - parse each message into the type of the handler's `message` kwarg
+        - invoke handler with resolved dependencies
+        """
+        assert self.handler_parameter_model, "handler parameter model not defined"
+        # `data` must be declared in the handler signature, but we manually provide the kwarg
```
Comment (Goldziher) [inline starlite/routes.py:343]: in this case you st make sure to validate this kwarg is in place

Comment (Goldziher) [reply starlite/routes.py:343]: So, look in the `Starlite` class itself, where the signature models are actually generated - this also needs to be updated.

File: starlite/connection.py (hunk 1 above)
Comment (peterschutt) [reply starlite/connection.py:84]: For sure, will incorporate when I next iterate on this.

File: starlite/connection.py [hunk 8]
```diff
@@ -74,3 +75,32 @@ def auth(self) -> Auth:
...
+        if self.application_state != WebSocketState.CONNECTED:
+            raise RuntimeError('WebSocket is not connected. Need to call "accept" first.')
```
[the remark this answers is not in the document]
Comment (peterschutt) [reply starlite/connection.py:86]: yep

File: starlite/handlers/__init__.py (hunk 2 above)
Comment (peterschutt) [reply starlite/handlers/__init__.py:21]: sounds good

File: starlite/handlers/websocket_message.py (hunk 5 above)
Comment (peterschutt) [reply starlite/handlers/websocket_message.py:51]: yep

File: starlite/router.py (hunk 6 above)
Comment (peterschutt) [reply starlite/router.py:106]: hah yeh they are big

File: starlite/routes.py (hunk 7 above)
Comment (peterschutt) [reply starlite/routes.py:343]: will check out and report back

File: starlite/connection.py (hunk 1 above)
Comment (peterschutt) [reply starlite/connection.py:84]: Went with `InternalServerException`

File: starlite/handlers/websocket_message.py (hunk 4 above)
Comment (peterschutt) [reply starlite/handlers/websocket_message.py:41]: have included in TODO

File: starlite/handlers/websocket_message.py (hunk 3 above)
Comment (peterschutt) [reply starlite/handlers/websocket_message.py:35]: I'll leave considering all the hooks that should exist as a TODO for now.

File: starlite/router.py (hunk 6 above)
Comment (peterschutt) [reply starlite/router.py:106]: I've done this one, I don't think I have the types right for route_handler_method_map() either, but I'll add a review for that.

**Comment by peterschutt** [comment]
Updated to keep in touch with main, still a long way off being an actual implementation, but have addressed most of the initial review points.

File: starlite/router.py [hunk 9]
```diff
@@ -86,11 +106,11 @@ def __init__(
             self.register(value=route_handler)
 
     @property
-    def route_handler_method_map(self) -> Dict[str, Union[WebsocketRouteHandler, Dict[HttpMethod, HTTPRouteHandler]]]:
+    def route_handler_method_map(self) -> Dict[str, Union[WebsocketRouteHandler, Dict["HttpMethod", HTTPRouteHandler]]]:
```
Comment (peterschutt) [inline starlite/router.py:109]: Types here should include WSMessageHandler.

File: starlite/connection.py [hunk 10]
```diff
@@ -74,3 +75,32 @@ def auth(self) -> Auth:
     @property
     def query_params(self) -> Dict[str, Any]:  # type: ignore[override]
         return parse_query_params(self)
+
```
Comment (Goldziher) [inline starlite/connection.py:78]: can you push the changes in this file into main? this is something we need to have in place regardless of this PR

**Pull request closed without merging**